File: clojure/meetup/meetup.clj
(ns meetup
  (:import (java.util GregorianCalendar)))

;; This exercise has a terrible API. Using keywords would've been a far better
;; choice than this combinatorial explosion of exported functions.
;; I guess this is really an exercise about metaprogramming and not dates?

(def ^:const ^:private weekdays
  {"sunday"    GregorianCalendar/SUNDAY
   "monday"    GregorianCalendar/MONDAY
   "tuesday"   GregorianCalendar/TUESDAY
   "wednesday" GregorianCalendar/WEDNESDAY
   "thursday"  GregorianCalendar/THURSDAY
   "friday"    GregorianCalendar/FRIDAY
   "saturday"  GregorianCalendar/SATURDAY})

(def ^:const ^:private DOW GregorianCalendar/DAY_OF_WEEK)

(def ^:const ^:private DOM GregorianCalendar/DAY_OF_MONTH)

(def ^:const ^:private ordinals
  {"first"  0
   "second" 1
   "third"  2
   "fourth" 3})

(defn- remove-day [s]
  (subs s 0 (- (count s) 3)))

(defn- set-weekday-forward! [cal weekday]
  (let [delta (- weekday (.get cal DOW))]
    (.add cal DOM (if (< delta 0) (+ delta 7) delta))))

(defn- set-weekday-reverse! [cal weekday]
  (let [delta (- weekday (.get cal DOW))]
    (.add cal DOM (if (> delta 0) (- delta 7) delta))))

(defn- set-end-of-month! [cal]
  (.set cal DOM (.getActualMaximum cal DOM)))

(defn- from-cal [cal]
  [(.get cal GregorianCalendar/YEAR)
   (inc (.get cal GregorianCalendar/MONTH))
   (.get cal DOM)])

(defn- def-cal-fun [name-str date & forms]
  (eval
   `(defn ~(symbol name-str) [month# year#]
      (from-cal (doto (GregorianCalendar. year# (dec month#) ~date) ~@forms)))))

;; first-sunday through fourth-saturday
(doseq [[ord-name ord] ordinals
        [weekday-name weekday] weekdays]
  (def-cal-fun (str ord-name "-" weekday-name) (inc (* 7 ord))
    `(set-weekday-forward! ~weekday)))

;; last-sunday through last-monday
(doseq [[weekday-name weekday] weekdays]
  (def-cal-fun (str "last-" weekday-name) 1
    `(set-end-of-month!)
    `(set-weekday-reverse! ~weekday)))

;; sunteenth through saturteenth
(doseq [[weekday-name weekday] weekdays]
  (def-cal-fun (str (remove-day weekday-name) "teenth") 13
    `(set-weekday-forward! ~weekday)))

File: elixir/meetup/meetup.exs
# This exercise has a terrible API. Using atoms would've been a far better
# choice than this combinatorial explosion of exported functions.
# I guess this is really an exercise about metaprogramming and not dates?
defmodule Meetup do
  # This is a much less terrible API
  def meetup_last(year, month, weekday) do
    start_day = :calendar.last_day_of_the_month(year, month) - 6
    meetup_nth(year, month, start_day, weekday)
  end
  def meetup_nth(year, month, start_day, weekday) do
    case weekday - :calendar.day_of_the_week(year, month, start_day) do
       n when n < 0 -> {year, month, start_day + n + 7}
       n            -> {year, month, start_day + n}
    end
  end

  # Here's the awful
  Enum.each([ {1, "mon", "monday"},
              {2, "tues", "tuesday"},
              {3, "wednes", "wednesday"},
              {4, "thurs", "thursday"},
              {5, "fri", "friday"},
              {6, "satur", "saturday"},
              {7, "sun", "sunday"} ],
    fn ({weekday, short, long}) ->
      Enum.each([first: 1, second: 8, third: 15, fourth: 22],
        fn ({ordinal, day}) ->
          def unquote(binary_to_atom("#{ordinal}_#{long}"))(month, year) do
            meetup_nth(year, month, unquote(day), unquote(weekday))
          end
        end)
      def unquote(binary_to_atom("#{short}teenth"))(month, year) do
            meetup_nth(year, month, 13, unquote(weekday))
      end
      def unquote(binary_to_atom("last_#{long}"))(month, year) do
            meetup_last(year, month, unquote(weekday))
      end
    end)
end
